controllers/virtualmachinegrouppublishrequest/virtualmachinegrouppublishrequest_controller.go

```diff
@@ -95,63 +110,288 @@ func NewReconciler(
+		}
+		if owned {
+			createdRequestsMap[request.Spec.Source.Name] = request
+			if conditions.IsTrue(&request, vmopv1.VirtualMachinePublishRequestConditionComplete) && request.Status.Ready {
+				completedRequestsSet.Insert(request.Spec.Source.Name)
+			} else {
+				pendingRequestsSet.Insert(request.Spec.Source.Name)
+			}
+		}
+	}
+
+	return createdRequestsMap, pendingRequestsSet, completedRequestsSet, nil
+}
+
+func (r *Reconciler) getSortedImageStatus(createdRequestsMap map[string]vmopv1.VirtualMachinePublishRequest) []vmopv1.VirtualMachineGroupPublishRequestImageStatus {
+	updatedImages := make([]vmopv1.VirtualMachineGroupPublishRequestImageStatus, 0, len(createdRequestsMap))
+	for _, createdRequest := range createdRequestsMap {
+		image := vmopv1.VirtualMachineGroupPublishRequestImageStatus{
+			Source:             createdRequest.Spec.Source.Name,
+			PublishRequestName: createdRequest.Name,
+			Conditions:         createdRequest.Status.Conditions,
+		}
+		if createdRequest.Status.ImageName != "" {
+			image.ImageName = createdRequest.Status.ImageName
+		}
+		updatedImages = append(updatedImages, image)
+	}
+
+	sort.Slice(updatedImages, func(i, j int) bool {
+		return updatedImages[i].Source < updatedImages[j].Source
+	})
+	return updatedImages
+}
+
+func (r *Reconciler) createNewVMPublishRequests(
+	ctx *pkgctx.VirtualMachineGroupPublishRequestContext,
+	createdRequestsMap map[string]vmopv1.VirtualMachinePublishRequest,
+	desiredRequestsSet sets.Set[string]) error {
+
+	createdRequestsSet := sets.KeySet(createdRequestsMap)
+	toBeCreatedRequestsSet := desiredRequestsSet.Difference(createdRequestsSet)
+	gvk, err := apiutil.GVKForObject(ctx.VMGroupPublishRequest, r.Scheme())
+	if err != nil {
+		return err
+	}
+
+	var errs []error
+	for toBeCreatedRequest := range toBeCreatedRequestsSet {
+		request := &vmopv1.VirtualMachinePublishRequest{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:    ctx.VMGroupPublishRequest.Namespace,
+				GenerateName: ctx.VMGroupPublishRequest.Name + "-",
+				OwnerReferences: []metav1.OwnerReference{
+					*metav1.NewControllerRef(ctx.VMGroupPublishRequest, gvk),
+				},
+			},
+			Spec: vmopv1.VirtualMachinePublishRequestSpec{
+				Source: vmopv1.VirtualMachinePublishRequestSource{
+					Name:       toBeCreatedRequest,
+					Kind:       reflect.TypeOf(vmopv1.VirtualMachine{}).Name(),
+					APIVersion: vmopv1.GroupVersion.String(),
+				},
+			},
+		}
+		if ctx.VMGroupPublishRequest.Spec.Target != "" {
+			request.Spec.Target = vmopv1.VirtualMachinePublishRequestTarget{
+				Location: vmopv1.VirtualMachinePublishRequestTargetLocation{
+					Name:       ctx.VMGroupPublishRequest.Spec.Target,
+					Kind:       reflect.TypeOf(imgregv1a1.ContentLibrary{}).Name(),
+					APIVersion: vmopv1.GroupVersion.String(),
+				},
+			}
+		}
+		if ctx.VMGroupPublishRequest.Spec.TTLSecondsAfterFinished != nil {
+			request.Spec.TTLSecondsAfterFinished = ctx.VMGroupPublishRequest.Spec.TTLSecondsAfterFinished
```

Comment:
This seems odd to me -  isn't the group's TTLSecondsAfterFinished just for the group itself. Each underlying publish is going to complete on its own pace.

Like if a one publish takes a long time, couldn't an already completed request delete itself, I guess causing this controller to create a new one?

Reply:
yeah, great catch! I was supposed to set the the TTL only when it is completed (if there is a TTL set). will fix. Thanks!